# <a id='toc1_'></a>[Data Consistency Checks](#toc0_)

**Table of contents**<a id='toc0_'></a>    
- [Data Consistency Checks](#toc1_)    
  - [I. Data Consistency Checks](#toc1_1_)    
    - [I.1. Dataframe df_prods](#toc1_1_1_)    
      - [I.1.1. Check for mixed-type data](#toc1_1_1_1_)    
      - [I.1.2. Check for missing values](#toc1_1_1_2_)    
      - [I.1.3. Address the missing values](#toc1_1_1_3_)    
      - [I.1.4. Check for duplicate values](#toc1_1_1_4_)    
    - [I.2. Dataframe df_ords](#toc1_1_2_)    
      - [I.2.1. Check for mixed-type data](#toc1_1_2_1_)    
      - [I.2.2. Check for missing values](#toc1_1_2_2_)    
      - [I.2.3. Address the missing values](#toc1_1_2_3_)    
      - [I.2.4. Check for duplicate values](#toc1_1_2_4_)    
  - [II. Data Export](#toc1_2_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

In [63]:
# import libraries
import numpy as np
import pandas as pd
import os

In [64]:
# create a path to the directory
path = r'C:\Users\Ansgar.S\Uyen\OneDrive\Documents\Data Immersion\Achievement IV - Python Fundamentals for Data Analysts\02-2023 Instacart Basket Analysis'

# import the 'products_wrangled.pkl' dataset
df_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'products_wrangled.pkl'))

# import the 'orders_wrangled.pkl' dataset
df_ords = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.pkl'))

## <a id='toc1_1_'></a>[I. Data Consistency Checks](#toc0_)

### <a id='toc1_1_1_'></a>[I.1. Dataframe df_prods](#toc0_)

#### <a id='toc1_1_1_1_'></a>[I.1.1. Check for mixed-type data](#toc0_)

In [65]:
# run descriptive statistics
print('Descriptive statistics:')
df_prods.describe().apply(lambda s: s.apply('{0:.2f}'.format))

Descriptive statistics:


prices
count  49693.00
mean       9.99
std      453.52
min        1.00
25%        4.10
50%        7.10
75%       11.20
max    99999.00

**Note**:
The maximum value of the prices column is 99,999 which is unusually high. The reason might be that it is a placeholder for missing values.

In [66]:
# check for mixed-type columns
for col in df_prods.columns.tolist():
    weird = (
        df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)
        ).any(axis = 1)
    if len (df_prods[weird]) >0:
        print(col, ': mixed types')
    else:
        print(col, ': consistent')

product_id : consistent
product_name : mixed types
aisle_id : consistent
department_id : consistent
prices : consistent


In [67]:
# check the current types of values in column 'product_name'
print('Current types of values in column product_name:')
df_prods['product_name'].apply(type).value_counts()

Current types of values in column product_name:


<class 'str'>      49677
<class 'float'>       16
Name: product_name, dtype: int64

In [68]:
# check a few values in column 'product_name'
print('Values in column product_name:')
df_prods['product_name'].sample(5)

Values in column product_name:


29476                                              Bowls
37599    Dead Sea Minerals Eucalyptus Triple Milled Soap
7234     Light & Fit Crunch Strawberry Cheesecake Yogurt
46201                              Liar's Dice Zinfandel
6682                                 California Apricots
Name: product_name, dtype: object

#### <a id='toc1_1_1_2_'></a>[I.1.2. Check for missing values](#toc0_)

In [69]:
# check for missing values
print('Missing values:')
df_prods.isnull().sum()

Missing values:


product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [70]:
# create a subset of records with missing values from column 'product_name'
df_prods_nan = df_prods[df_prods['product_name'].isnull() == True]

print('Records with missing values from column product_name:')
df_prods_nan

Records with missing values from column product_name:


product_id product_name aisle_id department_id  prices
33            34          NaN      121            14    12.2
68            69          NaN       26             7    11.8
115          116          NaN       93             3    10.8
261          262          NaN      110            13    12.1
525          525          NaN      109            11     1.2
1511        1511          NaN       84            16    14.3
1780        1780          NaN      126            11    12.3
2240        2240          NaN       52             1    14.2
2586        2586          NaN      104            13    12.4
3159        3159          NaN      126            11    13.1
3230        3230          NaN      120            16    14.4
3736        3736          NaN       41             8    14.8
4283        4283          NaN       77             7    14.4
4790        4790          NaN       91            16    14.5
38187      38183          NaN       39            12    20.9
40444      40440          NaN      120            16    14.8

#### <a id='toc1_1_1_3_'></a>[I.1.3. Address the missing values](#toc0_)

In [71]:
# check the number of rows and columns
print('Number of rows and columns:')
df_prods.shape

Number of rows and columns:


(49693, 5)

In [72]:
# create a new dataframe without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

print('df_prods without the missing values:')
df_prods_clean.head(5)

df_prods without the missing values:


product_id                                       product_name aisle_id  \
0          1                         Chocolate Sandwich Cookies       61   
1          2                                   All-Seasons Salt      104   
2          3               Robust Golden Unsweetened Oolong Tea       94   
3          4  Smart Ones Classic Favorites Mini Rigatoni Wit...       38   
4          5                          Green Chile Anytime Sauce        5   

  department_id  prices  
0            19     5.8  
1            13     9.3  
2             7     4.5  
3             1    10.5  
4            13     4.3

In [73]:
# count the number of rows and columns in df_prods_clean
print('Number of rows and columns in df_prods_clean:')
df_prods_clean.shape

Number of rows and columns in df_prods_clean:


(49677, 5)

In [74]:
# check the types of values in product_name column in df_prods_clean
print('Types of values in product_name column in df_prods_clean:')
df_prods_clean['product_name'].apply(type).value_counts()

Types of values in product_name column in df_prods_clean:


<class 'str'>    49677
Name: product_name, dtype: int64

#### <a id='toc1_1_1_4_'></a>[I.1.4. Check for duplicate values](#toc0_)

In [75]:
# create a subset of full duplicates in df_prods_clean
df_dupes = df_prods_clean[df_prods_clean.duplicated()]
print('Duplicates in df_prods_clean:')
df_dupes

Duplicates in df_prods_clean:


product_id                                       product_name aisle_id  \
462          462                  Fiber 4g Gummy Dietary Supplement       70   
18459      18458                                         Ranger IPA       27   
26810      26808               Black House Coffee Roasty Stout Beer       27   
35309      35306  Gluten Free Organic Peanut Butter & Chocolate ...      121   
35495      35491                            Adore Forever Body Wash      127   

      department_id  prices  
462              11     4.8  
18459             5     9.2  
26810             5    13.4  
35309            14     6.8  
35495            11     9.9

In [76]:
# check the number of rows in df_dupes
print('Number of rows and columns in df_dupes:')
df_dupes.shape

Number of rows and columns in df_dupes:


(5, 5)

In [77]:
# create a new dataframe without duplicates from df_prods_clean
df_prods_clean_no_dupes = df_prods_clean.drop_duplicates()

print('df_prods_clean without duplicates:') 
df_prods_clean_no_dupes.head(5)

df_prods_clean without duplicates:


product_id                                       product_name aisle_id  \
0          1                         Chocolate Sandwich Cookies       61   
1          2                                   All-Seasons Salt      104   
2          3               Robust Golden Unsweetened Oolong Tea       94   
3          4  Smart Ones Classic Favorites Mini Rigatoni Wit...       38   
4          5                          Green Chile Anytime Sauce        5   

  department_id  prices  
0            19     5.8  
1            13     9.3  
2             7     4.5  
3             1    10.5  
4            13     4.3

In [78]:
# count the number of rows and columns in df_prods_clean_no_dupes
print('Number of rows and columns in df_prods_clean_no_dupes:')
df_prods_clean_no_dupes.shape

Number of rows and columns in df_prods_clean_no_dupes:


(49672, 5)

### <a id='toc1_1_2_'></a>[I.2. Dataframe df_ords](#toc0_)

In [79]:
# run descriptive statistics
print('Descriptive statistics:')
df_ords.describe().apply(lambda s: s.apply('{0:.2f}'.format))

Descriptive statistics:


order_number orders_day_of_week order_hour_of_day days_since_prior_order
count   3421083.00         3421083.00        3421083.00             3214874.00
mean         17.15               2.78             13.45                  11.11
std          17.73               2.05              4.23                   9.21
min           1.00               0.00              0.00                   0.00
25%           5.00               1.00             10.00                   4.00
50%          11.00               3.00             13.00                   7.00
75%          23.00               5.00             16.00                  15.00
max         100.00               6.00             23.00                  30.00

**Observation**:
The count of the days_since_prior_order column is lower than other columns', which could indicate that some of the rows in the dataframe are missing values for this column.

#### <a id='toc1_1_2_1_'></a>[I.2.1. Check for mixed-type data](#toc0_)

In [80]:
# check for mixed-type columns
for col in df_ords.columns.tolist():
    weird = (
        df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)
        ).any(axis = 1)
    if len (df_ords[weird]) >0:
        print(col, ': mixed types')
    else:
        print(col, ': consistent')

order_id : consistent
user_id : consistent
order_number : consistent
orders_day_of_week : consistent
order_hour_of_day : consistent
days_since_prior_order : consistent


#### <a id='toc1_1_2_2_'></a>[I.2.2. Check for missing values](#toc0_)

In [81]:
# check for missing values
print('Missing values:')
df_ords.isnull().sum()

Missing values:


order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

**Observation**: The missing values in the days_since_prior_order column could be due to the fact that some orders are the first orders for the user, and thus there is no prior order to measure the days since. A solution would be to create a new column that indicates whether the customer is new or not by determining if the order is the first order for the user.

#### <a id='toc1_1_2_3_'></a>[I.2.3. Address the missing values](#toc0_)

In [82]:
# create a new column is_new_customer
df_ords['is_new_customer'] = df_ords['days_since_prior_order'].isna().apply(lambda x: 1 if x else 0)

print('df_ords with the new is_new_customer column:')
df_ords.head(10)

df_ords with the new is_new_customer column:


order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   
3  2254736       1             4                   4                  7   
4   431534       1             5                   4                 15   
5  3367565       1             6                   2                  7   
6   550135       1             7                   1                  9   
7  3108588       1             8                   1                 14   
8  2295261       1             9                   1                 16   
9  2550362       1            10                   4                  8   

   days_since_prior_order  is_new_customer  
0                     NaN                1  
1                    15.0                0  
2                    21.0                0  
3                    29.0                0  
4                    28.0                0  
5                    19.0                0  
6                    20.0                0  
7                    14.0                0  
8                     0.0                0  
9                    30.0                0

In [89]:
# run descriptive statistics
print('Descriptive statistics:')
df_ords.describe().apply(lambda s: s.apply('{0:.2f}'.format))

Descriptive statistics:


order_number orders_day_of_week order_hour_of_day  \
count   3421083.00         3421083.00        3421083.00   
mean         17.15               2.78             13.45   
std          17.73               2.05              4.23   
min           1.00               0.00              0.00   
25%           5.00               1.00             10.00   
50%          11.00               3.00             13.00   
75%          23.00               5.00             16.00   
max         100.00               6.00             23.00   

      days_since_prior_order is_new_customer  
count             3214874.00      3421083.00  
mean                   11.11            0.06  
std                     9.21            0.24  
min                     0.00            0.00  
25%                     4.00            0.00  
50%                     7.00            0.00  
75%                    15.00            0.00  
max                    30.00            1.00

**Observation**: This method is the most appropriate choice as it allows us to differentiate between new and existing customers without having to impute values in the days_since_prior_order column.

#### <a id='toc1_1_2_4_'></a>[I.2.4. Check for duplicate values](#toc0_)

In [90]:
# create a subset of full duplicates
df_ords_dupes = df_ords[df_ords.duplicated()]

print('Duplicates:')
df_ords_dupes

Duplicates:


Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, is_new_customer]
Index: []

**Observation**: 
There are no duplicates in df_ords.

## <a id='toc1_2_'></a>[II. Data Export](#toc0_)

In [85]:
# export df_prods_clean_no_dupes in .pkl format
df_prods_clean_no_dupes.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'products_cleaned.pkl'))

In [86]:
#export df_ords in .pkl format
df_ords.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_cleaned.pkl'))